Sources/ProcedureKit/AnyObserver.swift

```diff
@@ -62,6 +62,103 @@ class AnyObserverBox<Base: ProcedureObserver>: AnyObserverBox_<Base.Procedure> {
     }
 }
 
+/// Creates a ProcedureObserver that wraps another ProcedureObserver and transforms the input events from
+/// the input Procedure type ("R") to the wrapped ProcedureObserver's expected Procedure type ("O") (via a
+/// `procedureTransformBlock`).
+///
+/// This observer will fail at run-time if an event handler is called with a Procedure subclass that cannot
+/// be converted to the wrapped ProcedureObserver's expected Procedure type (via the transform block).
+///
+/// Failures are logged as warnings to the event's input Procedure.
+internal class TransformObserver<O: ProcedureProtocol, R: ProcedureProtocol>: ProcedureObserver {
+    private typealias Erased = AnyObserverBox_<O>
+    public typealias Procedure = R
+
+    private var wrapped: Erased
+    private var procedureTransformBlock: (R) -> O?
+    init<Base: ProcedureObserver>(base: Base, procedureTransformBlock: @escaping (R) -> O? = { return $0 as? O }) where O == Base.Procedure {
+        wrapped = AnyObserverBox(base)
+        self.procedureTransformBlock = procedureTransformBlock
+    }
+
+    private enum Event {
+        case didAttach
+        case willExecute
+        case didExecute
+        case didCancel
+        case willAdd
+        case didAdd
+        case willFinish
+        case didFinish
+
+        var string: String {
+            switch self {
+            case .didAttach: return "didAttach"
+            case .willExecute: return "willExecute"
+            case .didExecute: return "didExecute"
+            case .didCancel: return "didCancel"
+            case .willAdd: return "procedureWillAdd"
+            case .didAdd: return "procedureDidAdd"
+            case .willFinish: return "willFinish"
+            case .didFinish: return "didFinish"
+            }
+        }
+    }
+
+    private func typedProcedure(_ procedure: R, event: Event, logError: Bool = false) -> O? {
+        guard let typedProcedure = procedureTransformBlock(procedure) else {
+            procedure.log.warning(message: ("Observer will not receive event (\(event.string)). Unable to convert \(procedure) to the expected type \"\(String(describing: O.self))\""))
+            return nil
+        }
+        return typedProcedure
+    }
+
+    public func didAttach(to procedure: Procedure) {
```

Comment:
Hmm, does the compiler not complain about these public functions, on an internal class?